README
BitPay Key Utils version 2.0.0
==============================

This library provides utilities for use with the BitPay API. It enables creating keys, retrieving public keys, creating the SIN that is used in retrieving tokens from BitPay, and signing payloads for the `X-Signature` header in a BitPay API request.

DEPENDENCIES

	libssl-dev (OpenSSL 0.9.8+)
	SWIG (3.0.5+)
	Perl (5.18+)
	Make

INSTALLATION

Install from cpan by typing the following:

sudo cpanm BitPay::key_utils

or

Clone the github repository and type the following:

	perl Makefile.PL
	make
	make test
	make install

USAGE

Use the key utils by including `use BitPay::key_utils` in your project. This should give you access to the functions:

	my $pem = BitPay::key_utils::mygeneratePem(); #creates ECDSA Keys and sets the value of pem to the PEM encoding of the key
	my $pub = BitPay::key_utils::mygetPublicKeyFromPem($pem); #takes a pem string and sets the value of pubkey to the compressed public key extracted from the pem
	my $sin = BitPay::key_utils::mygenerateSinFromPem($pem); #gets the base58 unique identifier associated with the pem
	my $signature = BitPay::key_utils::mysignMessageWithPem($pem, "He's dead, Jim."); #sets signature to the signature of the sha256 of the message

DEV ENVIRONMENT

Clone the github repository.

To build and install this module type the following:

	./clean
	./build

API DOCUMENTATION

API Documentation is available on the [BitPay site](https://bitpay.com/api).

RUNNING THE TESTS

To run the tests type the following:

	perl test.pl


CHANGES
Revision history for Perl extension BitPay::KeyUtils.

2.0.0  Mon May 18 12:56:47 2015
	- original version;


test.pl
use BitPay::KeyUtils;
use strict;
use warnings;
use Test::More;

my $const_sin = "TeyN4LPrXiG5t2yuSamKqP3ynVk3F52iHrX";
my $const_pem = "-----BEGIN EC PRIVATE KEY-----\nMHQCAQEEICg7E4NN53YkaWuAwpoqjfAofjzKI7Jq1f532dX+0O6QoAcGBSuBBAAK\noUQDQgAEjZcNa6Kdz6GQwXcUD9iJ+t1tJZCx7hpqBuJV2/IrQBfue8jh8H7Q/4vX\nfAArmNMaGotTpjdnymWlMfszzXJhlw==\n-----END EC PRIVATE KEY-----\n\n";
my $const_pri = "283B13834DE77624696B80C29A2A8DF0287E3CCA23B26AD5FE77D9D5FED0EE90";
my $const_pub = "038D970D6BA29DCFA190C177140FD889FADD6D2590B1EE1A6A06E255DBF22B4017";

test_generate_pem();
test_get_sin_from_pem();
test_get_public_key_from_pem();
test_sign_message_with_pem();

sub test_generate_pem {
	my $pem = BitPay::KeyUtils::bpGeneratePem();
	ok( $pem =~ m/BEGIN EC PRIVATE KEY.*\n.*\n.*\n.*\n.*END EC PRIVATE KEY/, 'TEST PEM GENERATION');
}

sub test_get_sin_from_pem {
	my $sin = BitPay::KeyUtils::bpGenerateSinFromPem($const_pem);
	is( $sin, $const_sin, 'TEST SIN GENERATION');
}

sub test_get_public_key_from_pem {
	my $pub = BitPay::KeyUtils::bpGetPublicKeyFromPem($const_pem);
	is( $pub, $const_pub, 'TEST PUBLIC KEY GENERATION');
}

sub test_sign_message_with_pem {
	my $message = "We are the crystal gems";
	my $signature = BitPay::KeyUtils::bpSignMessageWithPem($const_pem, $message);
	ok( $signature =~ m/^[a-f0-9]+$/, 'TEST SIGNATURE IS HEX');
	ok( $signature =~ m/^304[4-6]022[0-1].*022[0-1]/, 'TEST SIGNATURE ENCODING SCHEME, BYTE LENGTH, & NEW SECTION LENGTHS');

	my $signature_length = length $signature;
	ok( $signature_length >= 140 && $signature_length <= 144, "TEST SIGNATURE LENGTH");
}

done_testing();


tests.c
#include "bitpay.h"

static void runPemTest();
static void runPublicKeyTest();
static void runSinTest();
static void runSignatureTest();

int main() {
    runPemTest();
    runPublicKeyTest();
    runSinTest();
    runSignatureTest();
    return 1;
}

static void runSignatureTest() {
    int signa;
    char *message = "https://test.bitpay.com/invoices{\"currency\":\"USD\",\"price\":100,\"token\":\"GVTANyBKSJRdSzy88P72H2LB7gky7o4J8bebVbVaF6pA\"}";
    char *pem = calloc(240, sizeof(char));
    char *signature = calloc(145, sizeof(char));
    char *actual_start = calloc(4, sizeof(char));
    char *expected_start;

    pem[240]='\0';
    generatePem(&pem);
    signa = signMessageWithPem(message, pem, &signature);
    if(signa == ERROR){
        printf("Signature Error.\n");
    };
    actual_start[4] = '\0';
    memcpy(actual_start, signature, 4);
    if(strlen(signature) == 140)
        expected_start = "3044";
    else if(strlen(signature) == 142)
        expected_start = "3045";
    else if(strlen(signature) == 144)
        expected_start = "3046";
    else
        printf("Signature is not a valid length\n");

    if(strcmp(actual_start, expected_start) == 0)
        printf(".");
    else
        printf("Signature test - Expected: %s, Actual: %s\n", expected_start, actual_start);
    printf("\n");

    free(pem);
    free(signature);
    free(actual_start);
}

static void runSinTest() {
    char *fixed_pem = "-----BEGIN EC PRIVATE KEY-----\nMHQCAQEEIOg1/L9j0a63o5mRtzmG6N3Cn76MDpbUd2ZYAy4kYmq1oAcGBSuBBAAK\noUQDQgAE2IauvNs634tRrspRnEMbv9dQ84xoqBFilQQkmhHZJde+/8VwpMQ4wIQP\nYB429LjWsy3VyOF8vUpUmIvx17g/7g==\n-----END EC PRIVATE KEY-----\n";
    int singood;
    char *expected_sin = "Tf41EHiUGugMMeAR35DXfUrfkjzwmvqRQkz";
    char *sin = calloc(35, sizeof(char));

    singood = generateSinFromPem(fixed_pem, &sin);
    if(singood == ERROR)
        printf("Sin Error\n");
    if(strcmp(expected_sin, sin) == 0)
        printf(".");
    else
        printf("Sin test - Expected: %s, Actual: %s\n", expected_sin, sin);

    free(sin);
}

static void runPublicKeyTest(){
    char *pub = malloc(67);
    char *fixed_pem = "-----BEGIN EC PRIVATE KEY-----\nMHQCAQEEIOg1/L9j0a63o5mRtzmG6N3Cn76MDpbUd2ZYAy4kYmq1oAcGBSuBBAAK\noUQDQgAE2IauvNs634tRrspRnEMbv9dQ84xoqBFilQQkmhHZJde+/8VwpMQ4wIQP\nYB429LjWsy3VyOF8vUpUmIvx17g/7g==\n-----END EC PRIVATE KEY-----\n";
    char *expected_pub = "02D886AEBCDB3ADF8B51AECA519C431BBFD750F38C68A811629504249A11D925D7";
    int pubgood = getPublicKeyFromPem(fixed_pem, &pub);
    if(pubgood == ERROR)
        printf("Error retrieving public key\n");
    if(strcmp(expected_pub, pub) == 0)
        printf(".");
    else
        printf("Public key test - Expected: %s, Actual: %s\n", expected_pub, pub);

    free(pub);
}

static void runPemTest() {
    char *expected_pem_S = "MHQCAQ";
    char *expected_pem_N = "SuBBAAK\n";
    char *actual_pem_S = calloc(7, sizeof(char));
    char *actual_pem_N = calloc(9, sizeof(char));
    char *pem = malloc(224);

    pem[239]='\0';
    generatePem(&pem);
    actual_pem_S[6] = '\0';
    actual_pem_N[8] = '\0';
    memcpy(actual_pem_S, pem+31, 6);
    memcpy(actual_pem_N, pem+88, 8);

    if(strcmp(expected_pem_S, actual_pem_S) == 0)
        printf(".");
    else
        printf("Pem test - Expected: %s, Actual: %s\n", expected_pem_S, actual_pem_S);
    if(strcmp(expected_pem_N, actual_pem_N) == 0)
        printf(".");
    else
        printf("Pem test - Expected: %s, Actual: %s\n", expected_pem_N, actual_pem_N);

    free(pem);
    free(actual_pem_S);
    free(actual_pem_N);
}


bitpay.h
/*
 * File:   bitpay.h
 * Author: paul
 *
 * Created on April 27, 2015, 3:44 PM
 */


#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <openssl/ec.h>
#include <openssl/bn.h>
#include <openssl/sha.h>
#include <openssl/ripemd.h>
#include <openssl/ecdsa.h>
#include <openssl/pem.h>

#define	NOERROR	0
#define	ERROR 	-1
#define SHA256_STRING 33
#define SHA256_HEX_STRING 65
#define RIPEMD_AND_PADDING 22
#define RIPEMD_AND_PADDING_STRING 23
#define RIPEMD_HEX 40
#define RIPEMD_HEX_STRING 41
#define RIPEMD_AND_PADDING_HEX_STRING 45
#define CHECKSUM 8
#define CHECKSUM_STRING 9
#define SIN 52
#define SIN_STRING 53

int generatePem(char **pem);
int generateSinFromPem(char *pem, char **sin);
int getPublicKeyFromPem(char *pemstring, char **pubkey);
int signMessageWithPem(char *pem, char *message, char **signature); 

static int createNewKey(EC_GROUP *group, EC_KEY *eckey);
static int createDataWithHexString(char *inputString, uint8_t **result);
static int base58encode(char *input, char *base58encode);
static int digestOfBytes(uint8_t *message, char **output, char *type, int inLength);


KeyUtils.pm
# This file was automatically generated by SWIG (http://www.swig.org).
# Version 3.0.5
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.

package BitPay::KeyUtils;
use base qw(Exporter);
use base qw(DynaLoader);
package BitPay::KeyUtilsc;
bootstrap BitPay::KeyUtils;
package BitPay::KeyUtils;
@EXPORT = qw();

# ---------- BASE METHODS -------------

package BitPay::KeyUtils;

sub TIEHASH {
    my ($classname,$obj) = @_;
    return bless $obj, $classname;
}

sub CLEAR { }

sub FIRSTKEY { }

sub NEXTKEY { }

sub FETCH {
    my ($self,$field) = @_;
    my $member_func = "swig_${field}_get";
    $self->$member_func();
}

sub STORE {
    my ($self,$field,$newval) = @_;
    my $member_func = "swig_${field}_set";
    $self->$member_func($newval);
}

sub this {
    my $ptr = shift;
    return tied(%$ptr);
}


# ------- FUNCTION WRAPPERS --------

package BitPay::KeyUtils;

*generatePem = *BitPay::KeyUtilsc::generatePem;
*generateSinFromPem = *BitPay::KeyUtilsc::generateSinFromPem;
*getPublicKeyFromPem = *BitPay::KeyUtilsc::getPublicKeyFromPem;
*signMessageWithPem = *BitPay::KeyUtilsc::signMessageWithPem;
*bpSignMessageWithPem = *BitPay::KeyUtilsc::bpSignMessageWithPem;
*bpGeneratePem = *BitPay::KeyUtilsc::bpGeneratePem;
*bpGenerateSinFromPem = *BitPay::KeyUtilsc::bpGenerateSinFromPem;
*bpGetPublicKeyFromPem = *BitPay::KeyUtilsc::bpGetPublicKeyFromPem;

# ------- VARIABLE STUBS --------

package BitPay::KeyUtils;

1;


key_utils.pm
# This file was automatically generated by SWIG (http://www.swig.org).
# Version 3.0.5
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.

package BitPay::key_utils;
use base qw(Exporter);
use base qw(DynaLoader);
package BitPay::key_utilsc;
bootstrap BitPay::key_utils;
package BitPay::key_utils;
@EXPORT = qw();

# ---------- BASE METHODS -------------

package BitPay::key_utils;

sub TIEHASH {
    my ($classname,$obj) = @_;
    return bless $obj, $classname;
}

sub CLEAR { }

sub FIRSTKEY { }

sub NEXTKEY { }

sub FETCH {
    my ($self,$field) = @_;
    my $member_func = "swig_${field}_get";
    $self->$member_func();
}

sub STORE {
    my ($self,$field,$newval) = @_;
    my $member_func = "swig_${field}_set";
    $self->$member_func($newval);
}

sub this {
    my $ptr = shift;
    return tied(%$ptr);
}


# ------- FUNCTION WRAPPERS --------

package BitPay::key_utils;

*generatePem = *BitPay::key_utilsc::generatePem;
*generateSinFromPem = *BitPay::key_utilsc::generateSinFromPem;
*getPublicKeyFromPem = *BitPay::key_utilsc::getPublicKeyFromPem;
*signMessageWithPem = *BitPay::key_utilsc::signMessageWithPem;
*mysignMessageWithPem = *BitPay::key_utilsc::mysignMessageWithPem;
*mygeneratePem = *BitPay::key_utilsc::mygeneratePem;
*mygenerateSinFromPem = *BitPay::key_utilsc::mygenerateSinFromPem;
*mygetPublicKeyFromPem = *BitPay::key_utilsc::mygetPublicKeyFromPem;

# ------- VARIABLE STUBS --------

package BitPay::key_utils;

1;


key_utils.c
#include <string.h>
#include "bitpay.h"

static int createNewKey(EC_GROUP *group, EC_KEY *eckey);
static int createDataWithHexString(char *inputString, uint8_t **result);
static int base58encode(char *input, char *base58encode);
static int digestOfBytes(uint8_t *message, char **output, char *type, int inLength);
static int toHexString(char *input, int inLength, char **output);

int generatePem(char **pem) {

    char *pemholder = calloc(224, sizeof(char));
    EC_KEY *eckey = NULL;

    BIO *out = BIO_new(BIO_s_mem());
    BUF_MEM *buf = NULL;
    EC_GROUP *group = NULL;
    
    group = EC_GROUP_new_by_curve_name(NID_secp256k1);
    buf = BUF_MEM_new();
    eckey = EC_KEY_new();

    createNewKey(group, eckey);
    
    EC_GROUP_free(group);

    PEM_write_bio_ECPrivateKey(out, eckey, NULL, NULL, 0, NULL, NULL);

    BIO_get_mem_ptr(out, &buf);

    memcpy(pemholder, buf->data, 223);

    if ( buf->data[219] == '\n') {
        pemholder[220] = '\0';       
        memcpy(*pem, pemholder, 221);
    } else if ( buf->data[221] == '\n') {
        pemholder[222] = '\0';       
        memcpy(*pem, pemholder, 223);
    } else {
        pemholder[223] = '\0';
        memcpy(*pem, pemholder, 224);
    }

    EC_KEY_free(eckey);
    BIO_free_all(out);

    return NOERROR;
};

static int createNewKey(EC_GROUP *group, EC_KEY *eckey) {

    int asn1Flag = OPENSSL_EC_NAMED_CURVE;
    int form = POINT_CONVERSION_UNCOMPRESSED;

    EC_GROUP_set_asn1_flag(group, asn1Flag);
    EC_GROUP_set_point_conversion_form(group, form);
    EC_KEY_set_group(eckey, group);

    int resultFromKeyGen = EC_KEY_generate_key(eckey);

    if (resultFromKeyGen != 1){
        return ERROR;
    }

    return NOERROR;
}

int generateSinFromPem(char *pem, char **sin) {
    
    char *pub =     calloc(67, sizeof(char));

    u_int8_t *outBytesPub = calloc(SHA256_STRING, sizeof(u_int8_t));
    u_int8_t *outBytesOfStep1 = calloc(SHA256_STRING, sizeof(u_int8_t));
    u_int8_t *outBytesOfStep3 = calloc(RIPEMD_AND_PADDING_STRING, sizeof(u_int8_t));
    u_int8_t *outBytesOfStep4a = calloc(SHA256_STRING, sizeof(u_int8_t));

    char *step1 =   calloc(SHA256_HEX_STRING, sizeof(char));
    char *step2 =   calloc(RIPEMD_HEX_STRING, sizeof(char));
    char *step3 =   calloc(RIPEMD_AND_PADDING_HEX_STRING, sizeof(char));
    char *step4a =  calloc(SHA256_HEX_STRING, sizeof(char));
    char *step4b =  calloc(SHA256_HEX_STRING, sizeof(char));
    char *step5 =   calloc(CHECKSUM_STRING, sizeof(char));
    char *step6 =   calloc(RIPEMD_AND_PADDING_HEX + CHECKSUM_STRING, sizeof(char));

    char *base58OfStep6 = calloc(SIN_STRING, sizeof(char));

    getPublicKeyFromPem(pem, &pub);
    pub[66] = '\0';

    createDataWithHexString(pub, &outBytesPub);
    digestOfBytes(outBytesPub, &step1, "sha256", SHA256_STRING);
    step1[64] = '\0';

    createDataWithHexString(step1, &outBytesOfStep1);
    digestOfBytes(outBytesOfStep1, &step2, "ripemd160", SHA256_DIGEST_LENGTH);
    step2[40] = '\0';
   
    memcpy(step3, "0F02", 4);
    memcpy(step3+4, step2, RIPEMD_HEX);
    step3[44] = '\0';

    createDataWithHexString(step3, &outBytesOfStep3);
    digestOfBytes(outBytesOfStep3, &step4a, "sha256", RIPEMD_AND_PADDING);
    step4a[64] = '\0';

    createDataWithHexString(step4a, &outBytesOfStep4a);
    digestOfBytes(outBytesOfStep4a, &step4b, "sha256", SHA256_DIGEST_LENGTH);
    step4b[64] = '\0';

    memcpy(step5, step4b, CHECKSUM);
    
    sprintf(step6, "%s%s", step3, step5);
    step6[RIPEMD_AND_PADDING_HEX + CHECKSUM] = '\0';

    base58encode(step6, base58OfStep6);
    base58OfStep6[SIN] = '\0'; 
    memcpy(*sin, base58OfStep6, SIN_STRING);

    free(pub);  
    free(step1);
    free(step2);
    free(step3);
    free(step4a);
    free(step4b);
    free(step5);
    free(step6);
    free(base58OfStep6);

    free(outBytesPub);
    free(outBytesOfStep1);
    free(outBytesOfStep3);
    free(outBytesOfStep4a);
    return NOERROR;
}

int getPublicKeyFromPem(char *pemstring, char **pubkey) {

    EC_KEY *eckey = NULL;
    EC_KEY *key = NULL;
    EC_POINT *pub_key = NULL;
    BIO *in = NULL;
    const EC_GROUP *group = NULL;
    char *hexPoint = NULL;
    char xval[65] = "";
    char yval[65] = "";
    char *oddNumbers = "13579BDF";

    BIGNUM start;
    const BIGNUM *res;
    BN_CTX *ctx;

    BN_init(&start);
    ctx = BN_CTX_new();

    res = &start;

    const char *cPem = pemstring;
    in = BIO_new(BIO_s_mem());
    BIO_puts(in, cPem);
    key = PEM_read_bio_ECPrivateKey(in, NULL, NULL, NULL);
    res = EC_KEY_get0_private_key(key);
    eckey = EC_KEY_new_by_curve_name(NID_secp256k1);
    group = EC_KEY_get0_group(eckey);
    pub_key = EC_POINT_new(group);

    EC_KEY_set_private_key(eckey, res);

    if (!EC_POINT_mul(group, pub_key, res, NULL, NULL, ctx)) {
        return ERROR;
    }

    EC_KEY_set_public_key(eckey, pub_key);

    hexPoint = EC_POINT_point2hex(group, pub_key, 4, ctx);

    char *hexPointxInit = hexPoint + 2;
    memcpy(xval, hexPointxInit, 64);

    char *hexPointyInit = hexPoint + 66;
    memcpy(yval, hexPointyInit, 64);

    char *lastY = hexPoint + 129;
    hexPoint[130] = '\0';

    char *buildCompPub = calloc(67, sizeof(char));

    if (strstr(oddNumbers, lastY) != NULL) {
        sprintf(buildCompPub, "03%s", xval);
        buildCompPub[66] = '\0';
        memcpy(*pubkey, buildCompPub, 67);
    } else {
        sprintf(buildCompPub, "02%s", xval);
        buildCompPub[66] = '\0';
        memcpy(*pubkey, buildCompPub, 67);
    }

    free(buildCompPub);
    BN_CTX_free(ctx);
    EC_KEY_free(eckey);
    EC_KEY_free(key);
    EC_POINT_free(pub_key);
    BIO_free(in);

    return NOERROR;
};

static int createDataWithHexString(char *inputString, uint8_t **result) {

    int i, o = 0;
    uint8_t outByte = 0;

    int inLength = strlen(inputString);

    uint8_t *outBytes = malloc(sizeof(uint8_t) * ((inLength / 2) + 1));

    for (i = 0; i < inLength; i++) {
        uint8_t c = inputString[i];
        int8_t value = -1;

        if      (c >= '0' && c <= '9') value =      (c - '0');
        else if (c >= 'A' && c <= 'F') value = 10 + (c - 'A');
        else if (c >= 'a' && c <= 'f') value = 10 + (c - 'a');

        if (value >= 0) {
            if (i % 2 == 1) {
                outBytes[o++] = (outByte << 4) | value;
                outByte = 0;
            } else {
                outByte = value;
            }

        } else {
            if (o != 0) break;
        }
    }

    memcpy(*result, outBytes, inLength/2);

    free(outBytes);

    return NOERROR;
}

static int base58encode(char *input, char *base58encode) {
    BIGNUM *bnfromhex = BN_new();
    BN_hex2bn(&bnfromhex, input);
    char *codeString = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

    char buildString[35];
    int lengthofstring = 0;
    int startat = 34;

    while(BN_is_zero(bnfromhex) != 1){
      int rem = BN_mod_word(bnfromhex, 58);
      buildString[startat] = codeString[rem];
      BN_div_word(bnfromhex, 58);
      lengthofstring++;
      startat--;
    }
    startat ++;
    
    int j = 0;
    int i;
    for (i = startat; i < lengthofstring; i++) {
      base58encode[j] = buildString[i];
      j++;
    }

    BN_free(bnfromhex);

    return NOERROR;
}

static int digestOfBytes(uint8_t *message, char **output, char *type, int inLength) {
    EVP_MD_CTX *mdctx;
    const EVP_MD *md;
    unsigned char md_value[EVP_MAX_MD_SIZE];
    unsigned int md_len, i;

    OpenSSL_add_all_digests();

    md = EVP_get_digestbyname(type);
    mdctx = EVP_MD_CTX_create();
    EVP_DigestInit_ex(mdctx, md, NULL);
    EVP_DigestUpdate(mdctx, message, inLength);
    EVP_DigestFinal_ex(mdctx, md_value, &md_len);
    EVP_MD_CTX_destroy(mdctx);

    char *digest = calloc((md_len*2) + 1, sizeof(char));
    for(i = 0; i < md_len; i++){
      sprintf(&digest[2*i], "%02x", md_value[i]);
    };
    digest[md_len * 2] = '\0';
    memcpy(*output, digest, strlen(digest));
    free(digest);
    /* Call this once before exit. */
    EVP_cleanup();
    return 0;
}

static int toHexString(char *input, int inLength, char **output) {
    
    uint8_t *byteData = (uint8_t*)malloc(inLength);
    memcpy(byteData, input, inLength);

    unsigned int i;
    char *digest = calloc((inLength*2) + 1, sizeof(char));

    for (i = 0; i < inLength; i++) {
        sprintf(&digest[2*i], "%02x", byteData[i]);
    }

    memcpy(*output, digest, strlen(digest));
    free(digest);
    return NOERROR;
}

int signMessageWithPem(char *message, char *pem, char **signature) {

    unsigned int meslen = strlen(message);
    unsigned char *messagebytes = calloc(meslen, sizeof(unsigned char));
    int derSigLen = 0;
    int i = 0;
    memcpy(messagebytes, message, meslen);

    EC_KEY *key = NULL;
    BIO *in = NULL;
    unsigned char *buffer = NULL;

    char *sha256ofMsg = calloc(SHA256_HEX_STRING, sizeof(char));
    unsigned char *outBytesOfsha256ofMsg = calloc(SHA256_STRING, sizeof(unsigned char));

    digestOfBytes(messagebytes, &sha256ofMsg, "sha256", meslen);
    sha256ofMsg[64] = '\0';
    createDataWithHexString(sha256ofMsg, &outBytesOfsha256ofMsg);
    
    in = BIO_new(BIO_s_mem());
    BIO_puts(in, pem);
    key = PEM_read_bio_ECPrivateKey(in, NULL, NULL, NULL);
    
    if(key == NULL){
       return ERROR;
    } 
    while(derSigLen < 70 && i < 10){
        i++;
        ECDSA_SIG *sig = ECDSA_do_sign((const unsigned char*)outBytesOfsha256ofMsg, SHA256_DIGEST_LENGTH, key);
        
        int verify = ECDSA_do_verify((const unsigned char*)outBytesOfsha256ofMsg, SHA256_DIGEST_LENGTH, sig, key);
        
        if(verify != 1) {
            return ERROR;
        }

        int buflen = ECDSA_size(key);
        buffer = OPENSSL_malloc(buflen);

        derSigLen = i2d_ECDSA_SIG(sig, &buffer);
    }
    if(i == 10)
        return ERROR;
    char *hexData = calloc(derSigLen, sizeof(char));
    memcpy(hexData, buffer-derSigLen, derSigLen);

    char *hexString = calloc(derSigLen*2+1, sizeof(char));

    toHexString(hexData, derSigLen, &hexString);
    hexString[derSigLen * 2] = '\0';
    
    memcpy(*signature, hexString, (derSigLen*2)+ 1);

    EC_KEY_free(key);

    BIO_free_all(in);
    free(sha256ofMsg);
    free(outBytesOfsha256ofMsg);
    free(hexData);
    free(hexString);

    return NOERROR;
}
